src/hal/src/device.rs

```diff
@@ -411,6 +411,24 @@ pub trait Device<B: Backend>: fmt::Debug + Any + Send + Sync {
         buf: &mut B::Buffer,
     ) -> Result<(), BindError>;
 
+    /// Create a dedicated allocation for a buffer.
+    ///
+    /// Returns a memory object that can't be used for binding any resources.
+    unsafe fn allocate_buffer_memory(
```

Comment:
To actually do that more efficiently, the function must return a view into memory with offset and size.

Reply:
The user is expected to still call `get_buffer_requirements` before doing this, so the size is known. And the offset is known to be zero. Is this not enough?

Reply:
Oh, so this will work only for dedicated allocations?

Reply:
I thought we want to integrate vma into vulkan backend this way

Reply:
I think what we can do with VMA is:
  1. make it a compile time optional dependency
  2. turn `Memory` into an enum, with one variant for normal allocations and another - for VMA-owned ones. The code dealing with memory would need to match the enum, but I don't think it's a problem.
  3. the user is then expected to treat the `Memory` as dedicated allocation

Does that sound reasonable?

Reply:
Then you will map ranges from two of those `VmaMemory` instances that actually comes from single `RealMemory` and violate vulkan usage.

Reply:
So the memory type exposed by the Vulkan backend would be:
```rust
enum Memory {
  Physical(vk::Memory),
  SubAllocated(vma::Allocation),
}
```
Mapping a `Memory::SubAllocated` will go though VMA and just give out a pointer without really doing any more Vulkan mapping, if I understand correctly, much like with Rendy-memory.

Reply:
If VMA has no limitations here. Rendy has to map all non-dedicated mappable memory persistently to allow this, which is not optimal. I wonder how VMA does that.

Reply:
> which is not optimal

Is it, really? My understanding is that it's both optimal and recommended for CPU-visible memory.

Reply:
It's not a problem when RAM is abundant, like on typical modern PC. But in other cases it may be not the best approach.

Reply:
Looks like it doesn't map memory by default on creation, unless there is a specific flag provided. Relevant docs - https://gpuopen-librariesandsdks.github.io/VulkanMemoryAllocator/html/memory_mapping.html
It confirms that VMA acts roughly like rendy-memory, and the approach this PR is taking should work.